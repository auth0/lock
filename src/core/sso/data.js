import webAPI from '../web_api';
import Cache from '../../utils/cache';

const cache = new Cache((...args) => webAPI.getSSOData(...args));

export function fetchSSOData(id, withAD, cb) {
  cache.get(id, withAD, cb);
}
