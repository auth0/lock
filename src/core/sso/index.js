import { Map } from 'immutable';

export function lastUsedConnection(m) {
  return m.getIn(['sso', 'lastUsedConnection']);
}

export function lastUsedUsername(m) {
  return m.getIn(['sso', 'lastUsedUsername'], '');
}
