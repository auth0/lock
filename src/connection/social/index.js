import Immutable from 'immutable';
import * as l from '../../core/index';
import { dataFns } from '../../utils/data_utils';

// TODO: Android version also has "unknonwn-social", "evernote" and
// "evernote-sandbox""evernote" in the list, considers "google-openid"
// to be enterprise and doesn't contain "salesforce-community". See
// https://github.com/auth0/Lock.Android/blob/98262cb7110e5d1c8a97e1129faf2621c1d8d111/lock/src/main/java/com/auth0/android/lock/utils/Strategies.java
export const STRATEGIES = {
  amazon: 'Amazon',
  aol: 'Aol',
  baidu: '百度',
  bitbucket: 'Bitbucket',
  box: 'Box',
  dropbox: 'Dropbox',
  dwolla: 'Dwolla',
  ebay: 'ebay',
  exact: 'Exact',
  facebook: 'Facebook',
  fitbit: 'Fitbit',
  github: 'GitHub',
  'google-openid': 'Google OpenId',
  'google-oauth2': 'Google',
  instagram: 'Instagram',
  linkedin: 'LinkedIn',
  miicard: 'miiCard',
  paypal: 'PayPal',
  'paypal-sandbox': 'PayPal Sandbox',
  planningcenter: 'Planning Center',
  renren: '人人',
  salesforce: 'Salesforce',
  'salesforce-community': 'Salesforce Community',
  'salesforce-sandbox': 'Salesforce (sandbox)',
  evernote: 'Evernote',
  'evernote-sandbox': 'Evernote (sandbox)',
  shopify: 'Shopify',
  soundcloud: 'Soundcloud',
  thecity: 'The City',
  'thecity-sandbox': 'The City (sandbox)',
  thirtysevensignals: '37 Signals',
  twitter: 'Twitter',
  vkontakte: 'vKontakte',
  windowslive: 'Microsoft Account',
  wordpress: 'Wordpress',
  yahoo: 'Yahoo!',
  yammer: 'Yammer',
  yandex: 'Yandex',
  weibo: '新浪微博'
};

const { get, tget, initNS } = dataFns(['social']);

export function initSocial(m, options) {
  return initNS(m, Immutable.fromJS(processSocialOptions(options)));
}

export function displayName(connection) {
  if (['oauth1', 'oauth2'].indexOf(connection.get('strategy')) !== -1) {
    return connection.get('name');
  }
  return STRATEGIES[connection.get('strategy')];
}

export function processSocialOptions(options) {
  const result = {};
  const { socialButtonStyle } = options;

  // TODO: emit warnings
  if (['big', 'small'].indexOf(socialButtonStyle) > -1) {
    result.socialButtonStyle = socialButtonStyle;
  }

  return result;
}

export function socialConnections(m) {
  return l.connections(m, 'social');
}

export function authButtonsTheme(m) {
  return l.ui.authButtonsTheme(m);
}

export function useBigButtons(m, notFoundLimit) {
  const style = tget(m, 'socialButtonStyle') || get(m, 'socialButtonStyle');
  return style ? style === 'big' : l.connections(m, 'social').count() <= notFoundLimit;
}
