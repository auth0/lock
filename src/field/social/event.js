import * as l from '../../core/index';

export function emitFederatedLoginEvent(lock, provider, isSignUp) {
  let prov;

  try {
    prov = provider.toJS();
  } catch (e) {
    prov = provider;
  }

  l.emitEvent(lock, 'federated login', {
    name: prov.name,
    strategy: prov.strategy,
    action: isSignUp ? 'signup' : 'signin'
  });
}
