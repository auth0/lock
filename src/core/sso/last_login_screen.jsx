import React from 'react';
import Screen from '../screen';
import QuickAuthPane from '../../ui/pane/quick_auth_pane';
import { logIn, skipQuickAuth } from '../../quick-auth/actions';
import { lastUsedConnection, lastUsedUsername } from './index';
import * as l from '../index';
import { renderSignedInConfirmation } from '../signed_in_confirmation';
import { STRATEGIES as SOCIAL_STRATEGIES } from '../../connection/social/index';
import { authButtonsTheme } from '../../connection/social/index';

// TODO: handle this from CSS
function icon(strategy) {
  if (SOCIAL_STRATEGIES[strategy]) return strategy;
  if (strategy === 'google-apps') return strategy;
  if (~['adfs', 'office365', 'waad'].indexOf(strategy)) return 'windows';
  return 'auth0';
}

const Component = ({ i18n, model }) => {
  const headerText = i18n.html('lastLoginInstructions') || null;
  const header = headerText && <p>{headerText}</p>;
  const theme = authButtonsTheme(model);
  const connectionName = lastUsedConnection(model).get('name');
  const buttonTheme = theme.get(connectionName);
  const primaryColor = buttonTheme && buttonTheme.get('primaryColor');
  const foregroundColor = buttonTheme && buttonTheme.get('foregroundColor');
  const buttonIcon = buttonTheme && buttonTheme.get('icon');

  const buttonClickHandler = () => {
    logIn(l.id(model), lastUsedConnection(model), lastUsedUsername(model));
  };

  return (
    <QuickAuthPane
      alternativeLabel={i18n.str('notYourAccountAction')}
      alternativeClickHandler={() => skipQuickAuth(l.id(model))}
      buttonLabel={lastUsedUsername(model)}
      buttonClickHandler={buttonClickHandler}
      header={header}
      strategy={icon(lastUsedConnection(model).get('strategy'))}
      buttonIcon={buttonIcon}
      primaryColor={primaryColor}
      foregroundColor={foregroundColor}
    />
  );
};

export default class LastLoginScreen extends Screen {
  constructor() {
    super('lastLogin');
  }

  renderAuxiliaryPane(lock) {
    return renderSignedInConfirmation(lock);
  }

  render() {
    return Component;
  }
}
