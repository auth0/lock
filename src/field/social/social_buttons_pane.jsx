import React from 'react';
import AuthButton from '../../ui/button/auth_button';
import * as l from '../../core/index';
import { logIn } from '../../quick-auth/actions';
import { displayName, socialConnections, authButtonsTheme } from '../../connection/social/index';

export default class SocialButtonsPane extends React.Component {

  render() {
    // TODO: i don't like that it receives the instructions tanslated
    // but it also takes the t fn
    const {
      bigButtons,
      instructions,
      labelFn,
      lock,
      showLoading,
      signUp
    } = this.props;

    const headerText = instructions || null;
    const header = headerText && <p>{headerText}</p>;

    const themes = authButtonsTheme(lock);

    const buttons = socialConnections(lock).map(x => {
      const buttonTheme = themes.get(x.get("name"));
      const connectionName = buttonTheme && buttonTheme.get("displayName");
      const primaryColor = buttonTheme && buttonTheme.get("primaryColor");
      const foregroundColor = buttonTheme && buttonTheme.get("foregroundColor");
      const icon = buttonTheme && buttonTheme.get("icon");

      return(<AuthButton
        isBig={bigButtons}
        key={x.get("name")}
        label={labelFn(signUp ? "signUpWithLabel" : "loginWithLabel", connectionName || displayName(x))}
        onClick={() => logIn(l.id(lock), x)}
        strategy={x.get("strategy")}
        primaryColor={primaryColor}
        foregroundColor={foregroundColor}
        icon={icon}
      />)
    });

    const loading = showLoading
      && <div className="auth0-loading-container">
           <div className="auth0-loading" />
         </div>;

    return (
      <div className="auth-lock-social-buttons-pane">
        {header}
        <div className="auth0-lock-social-buttons-container">{buttons}</div>
        {loading}
      </div>
    );
  }

}

SocialButtonsPane.propTypes = {
  bigButtons: React.PropTypes.bool.isRequired,
  instructions: React.PropTypes.any,
  labelFn: React.PropTypes.func.isRequired,
  lock: React.PropTypes.object.isRequired,
  showLoading: React.PropTypes.bool.isRequired,
  signUp: React.PropTypes.bool.isRequired
};

SocialButtonsPane.defaultProps = {
  showLoading: false
};
