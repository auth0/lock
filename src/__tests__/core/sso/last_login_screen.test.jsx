import React from 'react';
import { mount } from 'enzyme';
import Immutable from 'immutable';

import { expectComponent, extractPropsFromWrapper, mockComponent } from 'testUtils';

jest.mock('ui/pane/quick_auth_pane', () => mockComponent('quick_auth_pane'));

//there's a circular dependency with this module, so we need to mock it
jest.mock('engine/classic');

const getComponent = () => {
  const LastLoginScreen = require('core/sso/last_login_screen').default;
  const screen = new LastLoginScreen();
  return screen.render();
};

describe('LastLoginScreen', () => {
  beforeEach(() => {
    jest.resetModules();

    jest.mock('quick-auth/actions', () => ({
      logIn: jest.fn(),
      skipQuickAuth: jest.fn()
    }));

    jest.mock('core/index', () => ({
      id: () => 'id'
    }));

    jest.mock('core/sso/index', () => ({
      lastUsedConnection: () => ({
        get: () => 'lastUsedConnection'
      }),
      lastUsedUsername: () => 'lastUsedUsername'
    }));

    jest.mock('connection/social/index', () => ({
      STRATEGIES: {},
      authButtonsTheme: () => ({
        get: () => undefined
      })
    }));
  });
  const defaultProps = {
    i18n: {
      str: (...keys) => keys.join(','),
      group: (...keys) => keys.join(','),
      html: (...keys) => keys.join(',')
    },
    model: 'model'
  };
  it('renders correctly', () => {
    const Component = getComponent();
    expectComponent(<Component {...defaultProps} />).toMatchSnapshot();
  });
  it('renders with custom connection theme', () => {
    require('connection/social/index').authButtonsTheme = () => ({
      get: () =>
        Immutable.fromJS({
          primaryColor: 'primaryColor',
          foregroundColor: 'foregroundColor',
          icon: 'icon'
        })
    });
    const Component = getComponent();
    expectComponent(<Component {...defaultProps} />).toMatchSnapshot();
  });
  describe('renders correct icon', () => {
    const testStrategy = strategy => {
      it(`when strategy is ${strategy}`, () => {
        require('core/sso/index').lastUsedConnection = () =>
          Immutable.fromJS({
            strategy
          });
        const Component = getComponent();
        expectComponent(<Component {...defaultProps} />).toMatchSnapshot();
      });
    };
    const testStrategyName = 'this-strategy-exists';
    require('connection/social/index').STRATEGIES = {
      [testStrategyName]: 'Test Strategy'
    };
    const strategies = [
      testStrategyName,
      'google-apps',
      'adfs',
      'office365',
      'waad',
      'some-other-strategy'
    ].forEach(testStrategy);
  });
  it('calls logIn in the buttonClickHandler', () => {
    const Component = getComponent();
    const wrapper = mount(<Component {...defaultProps} />);
    const props = extractPropsFromWrapper(wrapper);
    props.buttonClickHandler();
    const { mock } = require('quick-auth/actions').logIn;
    expect(mock.calls.length).toBe(1);
    expect(mock.calls[0][0]).toBe('id');
    expect(mock.calls[0][1].get()).toBe('lastUsedConnection');
    expect(mock.calls[0][2]).toBe('lastUsedUsername');
  });
  it('calls skipQuickAuth in the alternativeClickHandler', () => {
    const Component = getComponent();
    const wrapper = mount(<Component {...defaultProps} />);
    const props = extractPropsFromWrapper(wrapper);
    props.alternativeClickHandler();
    const { mock } = require('quick-auth/actions').skipQuickAuth;
    expect(mock.calls.length).toBe(1);
    expect(mock.calls[0][0]).toBe('id');
  });
});
