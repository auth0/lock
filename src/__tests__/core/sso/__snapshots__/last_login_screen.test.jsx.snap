// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LastLoginScreen renders correct buttonLabel uses SOCIAL_STRATEGY mapping when there is not a lastUsedUsername 1`] = `
<div
  data-__type="quick_auth_pane"
  data-alternativeClickHandler={[Function]}
  data-alternativeLabel="notYourAccountAction"
  data-buttonClickHandler={[Function]}
  data-buttonIcon={undefined}
  data-buttonLabel="Twitter"
  data-foregroundColor={undefined}
  data-header={
    <p>
      lastLoginInstructions
    </p>
  }
  data-primaryColor={undefined}
  data-strategy="twitter"
/>
`;

exports[`LastLoginScreen renders correct buttonLabel uses lastUsedConnectionName when there is not a lastUsedUsername and no SOCIAL_STRATEGY mapping 1`] = `
<div
  data-__type="quick_auth_pane"
  data-alternativeClickHandler={[Function]}
  data-alternativeLabel="notYourAccountAction"
  data-buttonClickHandler={[Function]}
  data-buttonIcon={undefined}
  data-buttonLabel="lastUsedConnection"
  data-foregroundColor={undefined}
  data-header={
    <p>
      lastLoginInstructions
    </p>
  }
  data-primaryColor={undefined}
  data-strategy="auth0"
/>
`;

exports[`LastLoginScreen renders correct icon when strategy is adfs 1`] = `
<div
  data-__type="quick_auth_pane"
  data-alternativeClickHandler={[Function]}
  data-alternativeLabel="notYourAccountAction"
  data-buttonClickHandler={[Function]}
  data-buttonIcon={undefined}
  data-buttonLabel="lastUsedUsername"
  data-foregroundColor={undefined}
  data-header={
    <p>
      lastLoginInstructions
    </p>
  }
  data-primaryColor={undefined}
  data-strategy="windows"
/>
`;

exports[`LastLoginScreen renders correct icon when strategy is empty, use name instead 1`] = `
<div
  data-__type="quick_auth_pane"
  data-alternativeClickHandler={[Function]}
  data-alternativeLabel="notYourAccountAction"
  data-buttonClickHandler={[Function]}
  data-buttonIcon={undefined}
  data-buttonLabel="lastUsedUsername"
  data-foregroundColor={undefined}
  data-header={
    <p>
      lastLoginInstructions
    </p>
  }
  data-primaryColor={undefined}
  data-strategy="auth0"
/>
`;

exports[`LastLoginScreen renders correct icon when strategy is google-apps 1`] = `
<div
  data-__type="quick_auth_pane"
  data-alternativeClickHandler={[Function]}
  data-alternativeLabel="notYourAccountAction"
  data-buttonClickHandler={[Function]}
  data-buttonIcon={undefined}
  data-buttonLabel="lastUsedUsername"
  data-foregroundColor={undefined}
  data-header={
    <p>
      lastLoginInstructions
    </p>
  }
  data-primaryColor={undefined}
  data-strategy="google-apps"
/>
`;

exports[`LastLoginScreen renders correct icon when strategy is office365 1`] = `
<div
  data-__type="quick_auth_pane"
  data-alternativeClickHandler={[Function]}
  data-alternativeLabel="notYourAccountAction"
  data-buttonClickHandler={[Function]}
  data-buttonIcon={undefined}
  data-buttonLabel="lastUsedUsername"
  data-foregroundColor={undefined}
  data-header={
    <p>
      lastLoginInstructions
    </p>
  }
  data-primaryColor={undefined}
  data-strategy="windows"
/>
`;

exports[`LastLoginScreen renders correct icon when strategy is some-other-strategy 1`] = `
<div
  data-__type="quick_auth_pane"
  data-alternativeClickHandler={[Function]}
  data-alternativeLabel="notYourAccountAction"
  data-buttonClickHandler={[Function]}
  data-buttonIcon={undefined}
  data-buttonLabel="lastUsedUsername"
  data-foregroundColor={undefined}
  data-header={
    <p>
      lastLoginInstructions
    </p>
  }
  data-primaryColor={undefined}
  data-strategy="auth0"
/>
`;

exports[`LastLoginScreen renders correct icon when strategy is this-strategy-exists 1`] = `
<div
  data-__type="quick_auth_pane"
  data-alternativeClickHandler={[Function]}
  data-alternativeLabel="notYourAccountAction"
  data-buttonClickHandler={[Function]}
  data-buttonIcon={undefined}
  data-buttonLabel="lastUsedUsername"
  data-foregroundColor={undefined}
  data-header={
    <p>
      lastLoginInstructions
    </p>
  }
  data-primaryColor={undefined}
  data-strategy="auth0"
/>
`;

exports[`LastLoginScreen renders correct icon when strategy is waad 1`] = `
<div
  data-__type="quick_auth_pane"
  data-alternativeClickHandler={[Function]}
  data-alternativeLabel="notYourAccountAction"
  data-buttonClickHandler={[Function]}
  data-buttonIcon={undefined}
  data-buttonLabel="lastUsedUsername"
  data-foregroundColor={undefined}
  data-header={
    <p>
      lastLoginInstructions
    </p>
  }
  data-primaryColor={undefined}
  data-strategy="windows"
/>
`;

exports[`LastLoginScreen renders correctly 1`] = `
<div
  data-__type="quick_auth_pane"
  data-alternativeClickHandler={[Function]}
  data-alternativeLabel="notYourAccountAction"
  data-buttonClickHandler={[Function]}
  data-buttonIcon={undefined}
  data-buttonLabel="lastUsedUsername"
  data-foregroundColor={undefined}
  data-header={
    <p>
      lastLoginInstructions
    </p>
  }
  data-primaryColor={undefined}
  data-strategy="auth0"
/>
`;

exports[`LastLoginScreen renders with custom connection theme 1`] = `
<div
  data-__type="quick_auth_pane"
  data-alternativeClickHandler={[Function]}
  data-alternativeLabel="notYourAccountAction"
  data-buttonClickHandler={[Function]}
  data-buttonIcon="icon"
  data-buttonLabel="lastUsedUsername"
  data-foregroundColor="foregroundColor"
  data-header={
    <p>
      lastLoginInstructions
    </p>
  }
  data-primaryColor="primaryColor"
  data-strategy="auth0"
/>
`;
